src/System.IO.Pipelines/PipeAwaitable.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal struct PipeAwaitable
+    {
+        private static readonly Action _awaitableIsCompleted = () => { };
+        private static readonly Action _awaitableIsNotCompleted = () => { };
+
+        private int _cancelledState;
+        private Action _state;
+        private readonly IScheduler _scheduler;
+
+        public PipeAwaitable(IScheduler scheduler, bool completed)
+        {
+            _cancelledState = CancelledState.NotCancelled;
+            _state = completed ? _awaitableIsCompleted : _awaitableIsNotCompleted;
+            _scheduler = scheduler;
+        }
+
+        public void Resume()
+        {
+            var awaitableState = Interlocked.Exchange(
+                ref _state,
+                _awaitableIsCompleted);
+
+            if (!ReferenceEquals(awaitableState, _awaitableIsCompleted) &&
+                !ReferenceEquals(awaitableState, _awaitableIsNotCompleted))
+            {
+                _scheduler.Schedule(awaitableState);
+            }
+        }
+
+        public void Reset()
+        {
+            // Change the state from observed -> not cancelled. We only want to reset the cancelled state if it was observed
+            Interlocked.CompareExchange(ref _cancelledState,
+                CancelledState.NotCancelled,
+                CancelledState.CancellationObserved);
+
+            if (_cancelledState != CancelledState.CancellationRequested)
```

Comment:
Take return value from the `CompareExchange` and use that instead? Rather than re-reading the value you've just cache invalidated